### DISCUSSION TITLE
Python Manim NumberLine Scaling Syntax

### DISCUSSION THREAD

**Discussion by SuperUser-Vermeylen** [ManimCommunity/manim#2758] [Q&A] [answered]
Manim NumberLine object:
[NumberLine](https://docs.manim.community/en/stable/reference/manim.mobject.graphing.number_line.NumberLine.html)
Manim scale object:
[Scaling](https://docs.manim.community/en/stable/_modules/manim/mobject/graphing/scale.html)

```
        a = NumberLine(
            x_range=[-2.51, 2.49, 0.5],
            length=13,
            decimal_number_config={"num_decimal_places": 1},
            include_numbers=True,
            font_size = 50,
            scaling = LinearBase(10).inverse_function(_),
        )
```

I don't understand how to tell my numberline to scale the x_range using the inverse_function or function parameter.

If I pass an underscore to the inverse_function, it tries to dvide the underscore by 10 resulting in operator error.
If I pass a float to the inverse_function it says, "'float' object has no attribute 'function'." 

x_range is supposed to be the value that is being divded, but it doesn't get passed well to anything other than scaling = LinearBase(10) or scaling = LogBase(10).

What's the proper syntax to scale my number line using the function or inverse_function parameter of LinearBase?

**Comment by behackl** [comment, marked as answer]
Can you explain what exactly you would like to do? `LinearBase.inverse_function` is used internally for plotting, not for constructing a base that can be passed to the `scaling` keyword argument.

**Reply by SuperUser-Vermeylen** [reply]
What I'm trying to do is take the inverse of each number on the numberline and my initial attempt was to scale it using a lambda function (lambda x: 1/x).
I might be using it completely incorrectly and I may need to create custom labels, but I'm struggling to move on from my ignorance of how the scaling parameter works.

Are you saying NumberLine x_range values can't be scaled with LinearBase.inverse_function?

**Reply by behackl** [reply]
With `LinearBase`, you can apply a *linear* (fixed) scale factor to the numbers on your `NumberPlane`, like, for example, scaling all numbers by a factor of 10. If I understand correctly, you are looking for a custom scale base; it is not too difficult to implement one on your own, here is an example:

```py
from typing import Iterable

from manim import *
from manim.mobject.graphing.scale import _ScaleBase

class InverseScale(_ScaleBase):
    def function(self, value: float) -> float:
        return 1 / value
    
    def inverse_function(self, value: float) -> float:
        return 1 / value

    def get_custom_labels(self, val_range: Iterable[float], unit_decimal_places=None) -> Iterable[Mobject]:
        return [MathTex(r"\frac{1}{%s}" % int(self.inverse_function(v))) for v in val_range]


class CustomScale(Scene):
    def construct(self):
        nl = NumberLine(
            x_range=[1, 10],
            scaling=InverseScale(custom_labels=True),
            include_numbers=True,
        )
        self.add(nl)
```

produces

![CustomScale_ManimCE_v0 15 2](https://user-images.githubusercontent.com/11851593/173059027-e0ab71a3-0673-487b-b9c8-cdf38d5ca03b.png)

Is this what you are looking for?

**Reply by SuperUser-Vermeylen** [reply]
This is extremely helpful, but I do have a question.

1)  Whats the best way to handle zero? 1/0 should be "undef," but python is replacing it with "inf." I could easily just accept it, but if there is a way to just exclude the value from 1/0 or "inf" from the numberline, I would prefer that over "inf."


https://user-images.githubusercontent.com/79375349/173186468-4e032ffa-8849-48a6-bff6-b88342576e59.mp4


Thank you very much for this.

**Reply by behackl** [reply]
If you still use the `get_custom_labels` method of the scale, you should be able to special case of inf (which most likely is NumPy's infinity, `np.inf` or so). Glad this was helpful!
